src/Microsoft.IdentityModel.Tokens/RsaSecurityKey.cs

```diff
@@ -27,69 +27,79 @@
-            bool hasPublicKey = rsaParameters.Exponent != null && rsaParameters.Modulus != null;
-            if (!hasPrivateKey && !hasPublicKey)
+            HasPrivateKey = rsaParameters.D != null && rsaParameters.DP != null && rsaParameters.DQ != null && rsaParameters.P != null && rsaParameters.Q != null;
+            HasPublicKey = rsaParameters.Exponent != null && rsaParameters.Modulus != null;
+            if (!HasPrivateKey && !HasPublicKey)
             {
-                throw new ArgumentException("no public or private key material found");
+                throw LogHelper.LogException<ArgumentException>("No public or private key material found");
             }
 
-            _rsaParamaeters = rsaParameters;
+            Parameters = rsaParameters;
         }
 
-        public override int KeySize
+        public RsaSecurityKey(RSA rsa)
         {
-            get
+            if (rsa == null)
+                throw LogHelper.LogArgumentNullException("rsa");
+
+            RSACryptoServiceProvider rsaCsp = rsa as RSACryptoServiceProvider;
+            if (rsaCsp != null)
             {
-                if (HasPublicKey)
-                    return _rsaParamaeters.Modulus.Length * 8;
-                else if (HasPrivateKey)
-                    return _rsaParamaeters.D.Length * 8;
-                else
-                    return 0;
+                HasPrivateKey = !rsaCsp.PublicOnly;
             }
-        }
-
-        public override bool HasPrivateKey
-        {
-            get
+            else
             {
-                return !(_rsaParamaeters.D == null || _rsaParamaeters.DP == null || _rsaParamaeters.DQ == null || _rsaParamaeters.P == null || _rsaParamaeters.Q == null);
+                // fake signing to determine if the rsa instance has the private key or not is a costly operation especially in case of HSM. We return true by default in that case, it will fail later at the time of signing or decrypting.
+                HasPrivateKey = true;
             }
+            HasPublicKey = true;
+            Rsa = rsa;
         }
 
-        public override bool HasPublicKey
+        public override bool HasPrivateKey { get; }
+
+        public override bool HasPublicKey { get; }
+
+        public override int KeySize
         {
             get
             {
-                return !(_rsaParamaeters.Exponent == null || _rsaParamaeters.Modulus == null);
+                if (Rsa != null)
+                    return Rsa.KeySize;
+                if (HasPublicKey)
+                    return Parameters.Modulus.Length * 8;
+                else if (HasPrivateKey)
+                    return Parameters.D.Length * 8;
+                else
+                    return 0;
             }
         }
 
+        /// <summary>
+        /// <see cref="RSAParameters"/> used to initialize the key.
+        /// </summary>
+        public RSAParameters Parameters { get; private set; }
```

Comment:
Keeping this property is probably not necessary, as you can extract the RSA parameters via `Rsa.ExportParameters`.